### PR TITLE
more robust precursor purity (PR 3 of X)

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

preparation for NuXL

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/ID/PrecursorPurity.h [hunk 1]
```diff
@@ -41,7 +41,8 @@ namespace OpenMS
       double target_intensity = 0.0;
       double signal_proportion = 0.0;
       Size target_peak_count = 0;
-      Size residual_peak_count = 0;
+      Size interfering_peak_count = 0;
+      PeakSpectrum interfering_peaks; // interfering peaks
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/ID/PrecursorPurity.h]:
```suggestion
      PeakSpectrum interfering_peaks; // peaks left after precursor (isotopic) peaks have been removed
```

**Comment by jpfeuffer** [comment]
No changes in IsobaricQuantifier?? Would be cool to have. This is broken for years, especially for SPS MS3

**Comment by timosachsenberg** [comment]
Do you know what would need to be done in IsobaricQuantifier? I wrote this code a long time ago but think it just skips the spectra with missing precursor (e.g., if MS1 got filtered)

**Comment by jpfeuffer** [comment]
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/ANALYSIS/QUANTITATION/IsobaricChannelExtractor.cpp#L45

I think just not relying on a strict MS1-MS2-MS3 hierarchy would solve the problem. Instead always look for/prefer the precursor that is annotated at the spectrum (which you probably do in your code). Ideally you move from the current spectrum index up or down or create/assume an indexed mzml.

**Comment by timosachsenberg** [comment]
so https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/KERNEL/MSExperiment.cpp#L515 would not work?

**Comment by jpfeuffer** [comment]
It would. But it is not used currently. And if there is a PrecursorPurity class, why not use the same scoring mechanism here?

**Pull request merged**